File: src/test/java/com/github/thaynarasilvapinto/SimuladorBanco/OperacaoControllerTest.java
package com.github.thaynarasilvapinto.SimuladorBanco;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.thaynarasilvapinto.SimuladorBanco.controller.request.OperacaoDepositoRequest;
import com.github.thaynarasilvapinto.SimuladorBanco.controller.request.OperacaoSaqueRequest;
import com.github.thaynarasilvapinto.SimuladorBanco.controller.request.OperacaoTransferenciaRequest;
import com.github.thaynarasilvapinto.SimuladorBanco.controller.response.OperacaoResponse;
import com.github.thaynarasilvapinto.SimuladorBanco.domain.Cliente;
import com.github.thaynarasilvapinto.SimuladorBanco.domain.Conta;
import com.github.thaynarasilvapinto.SimuladorBanco.domain.Operacao;
import com.github.thaynarasilvapinto.SimuladorBanco.domain.TipoOperacao;
import com.github.thaynarasilvapinto.SimuladorBanco.services.ClienteService;
import com.github.thaynarasilvapinto.SimuladorBanco.services.ContaService;
import com.github.thaynarasilvapinto.SimuladorBanco.services.OperacaoService;
import com.google.gson.Gson;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;

import java.util.List;

import static org.hamcrest.CoreMatchers.notNullValue;
import static org.junit.Assert.assertEquals;
import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@RunWith(SpringRunner.class)
@SpringBootTest
@AutoConfigureMockMvc
public class OperacaoControllerTest {
    @Autowired
    private MockMvc mvc;
    @Autowired
    private ClienteService clienteService;
    @Autowired
    private ContaService contaService;
    @Autowired
    private OperacaoService operacaoService;
    private Gson gson;
    private ObjectMapper mapper;
    private Cliente joao;
    private Cliente maria;

    private Conta joaoConta;
    private Conta mariaConta;

    private Operacao operacaoDepositoJoao;
    private Operacao operacaoSaqueJoao;
    private Operacao operacaoTransferencia;

    @Before
    public void setup() {
        createClient();
        this.gson = new Gson();
        this.mapper = new ObjectMapper();

        this.operacaoDepositoJoao = new Operacao(joaoConta, joaoConta, 200.00, TipoOperacao.DEPOSITO);
        this.operacaoSaqueJoao = new Operacao(joaoConta, joaoConta, 100.00, TipoOperacao.SAQUE);
        this.operacaoTransferencia = new Operacao(joaoConta, mariaConta, 100, TipoOperacao.TRANSFERENCIA);
    }

    private void createClient() {
        joaoConta = new Conta(0.00);
        joao = new Cliente("Joao Operacao Test ClienteController", "15142542675", null);
        mariaConta = new Conta(0.00);
        maria = new Cliente("Maria Operacao Test ClienteController", "17708289667", null);

        joao = clienteService.insert(joao);
        joaoConta = contaService.insert(joaoConta);
        joao.setConta(joaoConta);
        clienteService.update(joao);

        maria = clienteService.insert(maria);
        mariaConta = contaService.insert(mariaConta);
        maria.setConta(mariaConta);
        clienteService.update(maria);
    }

    @After
    public void delete() {
        clienteService.delete(joao.getId());
        clienteService.delete(maria.getId());
        List<Operacao> extrato = operacaoService.findAllContaOrigem(joaoConta);
        for (int i = 0; i < extrato.size(); i++) {
            operacaoService.delete(extrato.get(i).getIdOperacao());
        }
        extrato = operacaoService.findAllContaOrigem(mariaConta);
        for (int i = 0; i < extrato.size(); i++) {
            operacaoService.delete(extrato.get(i).getIdOperacao());
        }
        contaService.delete(joaoConta.getId());
        contaService.delete(mariaConta.getId());
    }


    @Test
    public void deveRetornarAOperacaosBuscada() throws Exception {

        operacaoDepositoJoao = operacaoService.insert(operacaoDepositoJoao);
        joaoConta.deposito(operacaoDepositoJoao);
        contaService.update(joaoConta);

        String body = gson.toJson(operacaoDepositoJoao);

        this.mvc.perform(get("/operacao/{id}", operacaoDepositoJoao.getIdOperacao()))
                .andExpect(status().isOk())
                .andExpect(content().json(body));

    }

    @Test
    public void deveRealizarDeposito() throws Exception {
        OperacaoDepositoRequest operacaoDepositoRequest = new OperacaoDepositoRequest(500.00);
        String content = gson.toJson(operacaoDepositoRequest);
        this.mvc.perform(post("/conta/{id}/deposito", joaoConta.getId())
                .content(content)
                .contentType(APPLICATION_JSON_UTF8))
                .andExpect(status().isOk())
                .andExpect(content().contentType(APPLICATION_JSON_UTF8))
                .andExpect(jsonPath("$.idOperacao", notNullValue()))
                .andDo(mvcResult -> {
                    OperacaoResponse response
                            = mapper.readValue(mvcResult.getResponse().getContentAsString(), OperacaoResponse.class);
                    assertEquals(500, response.getContaOrigem().getSaldo(), 0.0001);
                });


    }

    @Test
    public void deveRealizarSaque() throws Exception {

        joaoConta.setSaldo(300);
        contaService.update(joaoConta);

        OperacaoSaqueRequest operacaoSaqueRequest = new OperacaoSaqueRequest(200.00);
        String content = gson.toJson(operacaoSaqueRequest);
        this.mvc.perform(post("/conta/{id}/saque", joaoConta.getId())
                .content(content)
                .contentType(APPLICATION_JSON_UTF8))
                .andExpect(status().isOk())
                .andExpect(content().contentType(APPLICATION_JSON_UTF8))
                .andExpect(jsonPath("$.idOperacao", notNullValue()))
                .andDo(mvcResult -> {
                    OperacaoResponse response
                            = mapper.readValue(mvcResult.getResponse().getContentAsString(), OperacaoResponse.class);
                    assertEquals(100, response.getContaOrigem().getSaldo(), 0.0001);
                });
    }

    @Test
    public void deveRealizarTransferencia() throws Exception {
        joaoConta.setSaldo(300);
        contaService.update(joaoConta);

        OperacaoTransferenciaRequest operacaoTransferenciaRequest =
                new OperacaoTransferenciaRequest(100, mariaConta.getId());
        String content = gson.toJson(operacaoTransferenciaRequest);

        this.mvc.perform(post("/conta/{id}/transferencia", joaoConta.getId())
                .content(content)
                .contentType(APPLICATION_JSON_UTF8))
                .andExpect(status().isOk())
                .andExpect(content().contentType(APPLICATION_JSON_UTF8))
                .andExpect(jsonPath("$.idOperacao", notNullValue()))
                .andDo(mvcResult -> {
                    OperacaoResponse response
                            = mapper.readValue(mvcResult.getResponse().getContentAsString(), OperacaoResponse.class);
                    assertEquals(200, response.getContaOrigem().getSaldo(), 0.0001);
                    assertEquals(100, response.getContaDestino().getSaldo(), 0.0001);
                });


    }

}

File: src/test/java/com/github/thaynarasilvapinto/simuladorbanco/OperacaoControllerTest.kt
package com.github.thaynarasilvapinto.simuladorbanco

//package com.github.thaynarasilvapinto.simuladorbanco;

import com.fasterxml.jackson.databind.ObjectMapper
import com.github.thaynarasilvapinto.simuladorbanco.controller.request.OperacaoRequest
import com.github.thaynarasilvapinto.simuladorbanco.controller.response.DepositoResponse
import com.github.thaynarasilvapinto.simuladorbanco.controller.response.SaqueResponse
import com.github.thaynarasilvapinto.simuladorbanco.controller.response.TransferenciaResponse
import com.github.thaynarasilvapinto.simuladorbanco.domain.Cliente
import com.github.thaynarasilvapinto.simuladorbanco.domain.Conta
import com.github.thaynarasilvapinto.simuladorbanco.domain.Operacao
import com.github.thaynarasilvapinto.simuladorbanco.services.ClienteService
import com.github.thaynarasilvapinto.simuladorbanco.services.ContaService
import com.github.thaynarasilvapinto.simuladorbanco.services.OperacaoService
import com.google.gson.Gson
import org.hamcrest.CoreMatchers.notNullValue
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType.APPLICATION_JSON_UTF8
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*

@RunWith(SpringRunner::class)
@SpringBootTest
@AutoConfigureMockMvc
class OperacaoControllerTest{
    @Autowired
    private lateinit var mvc: MockMvc
    @Autowired
    private lateinit var clienteService: ClienteService
    @Autowired
    private lateinit var contaService: ContaService
    @Autowired
    private lateinit var operacaoService: OperacaoService
    private lateinit var gson: Gson
    private lateinit var joao: Cliente
    private lateinit var maria: Cliente

    private lateinit var joaoConta: Conta
    private lateinit var mariaConta: Conta

    private lateinit var operacaoDepositoJoao: Operacao
    private lateinit var operacaoSaqueJoao: Operacao
    private lateinit var operacaoTransferencia: Operacao

    @Before
    fun setup() {
        createClient()
        this.gson = Gson()

        this.operacaoDepositoJoao = Operacao(contaOrigem = joaoConta, contaDestino = joaoConta, valorOperacao = 200.00, tipoOperacao = Operacao.TipoOperacao.DEPOSITO)
        this.operacaoSaqueJoao = Operacao(contaOrigem = joaoConta, contaDestino = joaoConta, valorOperacao = 100.00, tipoOperacao = Operacao.TipoOperacao.SAQUE)
        this.operacaoTransferencia = Operacao(contaOrigem = joaoConta, contaDestino = mariaConta, valorOperacao = 100.00, tipoOperacao = Operacao.TipoOperacao.TRANSFERENCIA)
    }

    private fun createClient() {
        joaoConta = Conta(saldo = 0.00)
        joaoConta = contaService.insert(joaoConta)
        joao = Cliente(nome = "Joao Operacao Test ClienteController", cpf = "151.425.426-75", conta = joaoConta)
        mariaConta = Conta(saldo = 0.00)
        mariaConta = contaService.insert(mariaConta)
        maria = Cliente(nome = "Maria Operacao Test ClienteController", cpf = "177.082.896-67", conta = mariaConta)
        joao = clienteService.insert(joao)
        maria = clienteService.insert(maria)
    }

    @After
    fun delete() {
        clienteService.delete(joao.id)
        clienteService.delete(maria.id)
        var extrato = operacaoService.findAllContaOrigem(joaoConta)
        for (i in extrato.indices) {
            operacaoService.delete(extrato[i].idOperacao)
        }
        extrato = operacaoService.findAllContaOrigem(mariaConta)
        for (i in extrato.indices) {
            operacaoService.delete(extrato[i].idOperacao)
        }
        contaService.delete(joaoConta.id)
        contaService.delete(mariaConta.id)
    }

    @Test
    @Throws(Exception::class)
    fun `deve realizar deposito`() {
        val operacaoDepositoRequest = OperacaoRequest(valorOperacao = 500.00, contaDestino = null)
        val content = gson.toJson(operacaoDepositoRequest)
        this.mvc.perform(post("/conta/{id}/deposito", joaoConta.id)
                .content(content)
                .contentType(APPLICATION_JSON_UTF8))
                .andExpect(status().isOk)
                .andExpect(content().contentType(APPLICATION_JSON_UTF8))
                .andExpect(jsonPath("$.idOperacao", notNullValue()))
    }

    @Test
    @Throws(Exception::class)
    fun ` deve realizar saque`() {

        joaoConta.saldo = 300.00
        contaService.update(joaoConta)

        val operacaoSaqueRequest = OperacaoRequest(valorOperacao = 200.00, contaDestino = null)
        val content = gson.toJson(operacaoSaqueRequest)
        this.mvc.perform(post("/conta/{id}/saque", joaoConta.id)
                .content(content)
                .contentType(APPLICATION_JSON_UTF8))
                .andExpect(status().isOk)
                .andExpect(content().contentType(APPLICATION_JSON_UTF8))
                .andExpect(jsonPath("$.idOperacao", notNullValue()))
    }

    @Test
    @Throws(Exception::class)
    fun `deve realizar transferencia`() {
        joaoConta.saldo = 300.00
        contaService.update(joaoConta)

        val operacaoTransferenciaRequest = OperacaoRequest(valorOperacao = 100.00, contaDestino = mariaConta.id)
        val content = gson.toJson(operacaoTransferenciaRequest)

        this.mvc.perform(post("/conta/{id}/transferencia", joaoConta.id)
                .content(content)
                .contentType(APPLICATION_JSON_UTF8))
                .andExpect(status().isOk)
                .andExpect(content().contentType(APPLICATION_JSON_UTF8))
                .andExpect(jsonPath("$.idOperacao", notNullValue()))
    }

    @Test
    @Throws(Exception::class)
    fun `nao deve realizar deposito`() {
        val operacaoDepositoRequest = OperacaoRequest(valorOperacao = -500.00, contaDestino = null)
        val content = gson.toJson(operacaoDepositoRequest)
        this.mvc.perform(post("/conta/{id}/deposito", joaoConta.id)
                .content(content)
                .contentType(APPLICATION_JSON_UTF8))
                .andExpect(status().isBadRequest)
    }

    @Test
    @Throws(Exception::class)
    fun `nao deve realizar saque`() {
        val operacaoSaqueRequest = OperacaoRequest(valorOperacao = -200.00, contaDestino = null)
        val content = gson.toJson(operacaoSaqueRequest)
        this.mvc.perform(post("/conta/{id}/saque", joaoConta.id)
                .content(content)
                .contentType(APPLICATION_JSON_UTF8))
                .andExpect(status().isBadRequest)
    }

    @Test
    @Throws(Exception::class)
    fun `nao deve trealizar transfefrencia de valor negativo`() {
        val operacaoTransferenciaRequest = OperacaoRequest(valorOperacao = -100.00, contaDestino = mariaConta.id)
        val content = gson.toJson(operacaoTransferenciaRequest)

        this.mvc.perform(post("/conta/{id}/transferencia", joaoConta.id)
                .content(content)
                .contentType(APPLICATION_JSON_UTF8))
                .andExpect(status().isBadRequest)
    }

    @Test
    @Throws(Exception::class)
    fun `nao deve trealizar transfefrencia para mesmo conta`() {
        joaoConta.saldo = 300.00
        contaService.update(joaoConta)
        val operacaoTransferenciaRequest = OperacaoRequest(valorOperacao = 300.00, contaDestino = joaoConta.id)
        val content = gson.toJson(operacaoTransferenciaRequest)
        this.mvc.perform(post("/conta/{id}/transferencia", joaoConta.id)
                .content(content)
                .contentType(APPLICATION_JSON_UTF8))
                .andExpect(status().isUnprocessableEntity)
    }

}
